staging/src/k8s.io/component-base/metrics/registry.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"github.com/blang/semver"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	apimachineryversion "k8s.io/apimachinery/pkg/version"
+	"sync"
+)
+
+var globalRegistryFactory = metricsRegistryFactory{
+	globalRegistry: &noopKubeRegistry{},
+}
+
+// NewKubeRegistry creates a new kubernetes metric registry, loading in the kubernetes
+// version information available to the binary.
+func (r metricsRegistryFactory) newKubeRegistry() KubeRegistry {
+	if r.kubeVersion == nil {
+		return noopKubeRegistry{}
+	}
+	return newKubeRegistry(r.kubeVersion)
+}
+
+type metricsRegistryFactory struct {
+	globalRegistry KubeRegistry
+	kubeVersion    *apimachineryversion.Info
+	setVersionOnce sync.Once
+}
+
+// KubeRegistry is an interface which implements a subset of prometheus.Registerer and
+// prometheus.Gatherer interfaces
+type KubeRegistry interface {
+	Register(KubeCollector) error
+	MustRegister(...KubeCollector)
+	Unregister(KubeCollector) bool
+	Gather() ([]*dto.MetricFamily, error)
+}
+
+// kubeRegistry is a wrapper around a prometheus registry-type object. Upon initialization
+// the kubernetes binary version information is loaded into the registry object, so that
+// automatic behavior can be configured for metric versioning.
+type kubeRegistry struct {
+	PromRegistry
+	version semver.Version
+}
+
+// SetRegistryFactoryVersion sets the kubernetes version information for all
+// subsequent metrics registry initializations. Only the first call has an effect.
+// If a version is not set, then metrics registry creation will no-opt
+func SetRegistryFactoryVersion(ver *apimachineryversion.Info) {
```

Comment:
it overrides the registry. This means that everything registered before is gone?

Reply:
That's a good point. I have an idea for deferring registration until the version is actually set. It's in the newest changeset, would appreciate your opinions on it :)